src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,476 @@
+            writer.WriteKey("text");
+            WriteSourceText(writer, syntaxTree.GetText());
+            writer.WriteKey("parseOptions");
+            WriteParseOptions(writer, syntaxTree.Options);
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteSourceText(JsonWriter writer, SourceText? sourceText)
+        {
+            if (sourceText is null)
+            {
+                return;
+            }
+
+            writer.WriteObjectStart();
+            WriteByteArrayValue(writer, "checksum", sourceText.GetChecksum());
+            writer.Write("checksumAlgorithm", sourceText.ChecksumAlgorithm);
+            writer.Write("encoding", sourceText.Encoding?.EncodingName);
+            writer.WriteObjectEnd();
+        }
+
+        internal void WriteMetadataReference(JsonWriter writer, MetadataReference reference)
+        {
+            writer.WriteObjectStart();
+            if (reference is PortableExecutableReference peReference)
+            {
+                ModuleMetadata moduleMetadata;
+                switch (peReference.GetMetadata())
+                {
+                    case AssemblyMetadata assemblyMetadata:
+                        {
+                            if (assemblyMetadata.GetModules() is { Length: 1 } modules)
+                            {
+                                moduleMetadata = modules[0];
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException();
+                            }
+                        }
+                        break;
+                    case ModuleMetadata m:
+                        moduleMetadata = m;
+                        break;
+                    default:
+                        throw new InvalidOperationException();
+                }
+
+                // The path of a reference, unlike the path of a file, does not contribute to the output
+                // of the copmilation. Only the MVID, name and version contribute here hence the file path
+                // is deliberately omitted here.
+                if (moduleMetadata.GetMetadataReader() is { IsAssembly: true } peReader)
+                {
+                    var assemblyDef = peReader.GetAssemblyDefinition();
+                    writer.Write("name", peReader.GetString(assemblyDef.Name));
+                    writer.Write("version", assemblyDef.Version.ToString());
+                    WriteByteArrayValue(writer, "publicKey", peReader.GetBlobBytes(assemblyDef.PublicKey).AsSpan());
+                }
+
+                writer.Write("mvid", moduleMetadata.GetModuleVersionId().ToString());
+                writer.WriteKey("properties");
+                writeMetadataReferenceProperties(writer, reference.Properties);
+            }
+            else
+            {
+                throw new InvalidOperationException();
```

Comment:
Shouldn't we also be able to write the key of a compilation reference? This is the case when the IDE creates the compilation, right? Is that a TODO here or requires some design consideration to know what to do here?

Reply:
this is an interesting case.  would be happy if you solved it.  if you don't though, we'll just filter out those sorts of metadata references and then mix-in our own checksums for them into the final result.